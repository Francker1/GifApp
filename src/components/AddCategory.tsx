
import React, { useState } from 'react';
import PropTypes from 'prop-types';

type FormElement = React.FormEvent<HTMLFormElement>;

interface IMyProps {
    setCategories:any,
}


export const AddCategory:React.FC<IMyProps> = ( { setCategories }) => {

    const [inputValue, setInputValue] = useState("");

    const handleInputChange = (e:any) => {

        setInputValue(e.target.value)
    }

    const handleSubmit = (e:FormElement) => {

        e.preventDefault();

        if( inputValue.trim().length > 2 ){

            //si quisiera acumular debajo los elementos buscados:
            // setCategories( (cats:any) => [ inputValue, ...cats, ] );
            setCategories( () => [ inputValue ] );
            setInputValue("");
        }

    }

    return (
        <form onSubmit={ handleSubmit }>
            <input 
                type="text"
                value={inputValue}
                onChange={ handleInputChange }
            />
        </form>
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}
