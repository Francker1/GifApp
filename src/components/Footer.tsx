import React from 'react';


export const Footer = () => {
    return (
        <div>
            <hr />
            <p>Copyright &copy; √çtalo Franco</p>
        </div>
    )
}

