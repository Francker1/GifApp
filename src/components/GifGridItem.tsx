import React from 'react'

interface IMyProps {
    props?: any,
    id: string,
    title: string,
    url?: string,
    className: string
}

export const GifGridItem:React.FC<IMyProps> = ({ id, title, url }) => {
    
    return (
        <div className="card animate__animated animate__zoomIn">
            <img src={url} alt={title}/>
            <p>{title}</p>
            <p>ID: {id}</p>
        </div>
    )
}

