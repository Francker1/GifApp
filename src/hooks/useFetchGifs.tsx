import { useState, useEffect } from "react";
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = (category: string) => {

    const [state, setState] = useState<any>({
        data: [],
        loading: true,
    });

    useEffect(() =>{
        
        getGifs(category)
            .then( imgs => {

                //simulate resolve data from api with setTimeout
                setState({
                    data: imgs,
                    loading: false,
                }); 
            });
            
    }, [category])

    return state;
}