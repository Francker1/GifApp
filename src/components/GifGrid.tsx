
import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

interface IMyProps {
    category:string,
    className:string,
    data:[],
    loading:boolean,
}

export const GifGrid:React.FC<IMyProps> = ( {category} ) => {

    // const [images, setImages] = useState([]);

    const { data:images, loading } = useFetchGifs(category);

    return (
        <>
            <h3>{category}</h3>

            { loading && <p>Loading...</p> }
            
            <div className="card-grid">
                
                {
                    images.map( (img:any) => (                    
                        <GifGridItem 
                            key={img.id}
                            { ...img }
                        />
                    ))
                }
                
            </div>
        </>
    )
}

GifGrid.propTypes = {

    category: PropTypes.string.isRequired,
}