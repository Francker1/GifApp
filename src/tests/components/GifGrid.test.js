import React from 'react';
import "@testing-library/jest-dom";
import { shallow } from "enzyme";
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock("../../hooks/useFetchGifs");

describe('Tests with <GifGrid /> component', () => {

    const category = "Borat";
    
   
    test("should render GifrGrid component correctly", () => {

        useFetchGifs.mockReturnValue({
          data: [],
          loading: true,
        });

        const wrapper = shallow(<GifGrid category={category} />);
        expect(wrapper).toMatchSnapshot();
    });

    test("should show items when load images with useFetchGifs", () => {

        const gifs = [
          {
            id: "12345",
            url: "http://localhost/gif.jpg",
            title: "Random thing",
          },
          {
            id: "qwerty",
            url: "http://localhost/gif.jpg",
            title: "Random thing",
          },
        ];

        useFetchGifs.mockReturnValue({
          data: gifs,
          loading: false,
        });

        const wrapper = shallow(<GifGrid category={category} />);
        
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find("p").exists()).toBe(false);
        expect(wrapper.find("GifGridItem").length).toBe(gifs.length);
    });
});
