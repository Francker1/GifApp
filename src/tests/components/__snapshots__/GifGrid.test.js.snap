// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests with <GifGrid /> component should render GifrGrid component correctly 1`] = `
<Fragment>
  <h3>
    Borat
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`Tests with <GifGrid /> component should show items when load images with useFetchGifs 1`] = `
<Fragment>
  <h3>
    Borat
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="12345"
      key="12345"
      title="Random thing"
      url="http://localhost/gif.jpg"
    />
    <GifGridItem
      id="qwerty"
      key="qwerty"
      title="Random thing"
      url="http://localhost/gif.jpg"
    />
  </div>
</Fragment>
`;
